Start and stop TransportInfrastructure

SzymonPobiega opened the pull request:
Ability to perform, as a transport, actions before starting an endpoint and after stopping it (e.g. initialize and close connection to the broker).

The scenario where I discovered this missing extension point was the EventStore transport. ES uses "intelligent" thread-safe connections. Such connection should be initialized when starting the endpoint and closed when stopping it.

timbussmann commented:
@SzymonPobiega can you give some more context on why this is needed and labeled as v6?

bording commented:
I'm not sure I see the benefit of this. `TransportInfrastructure` gets created when `TransportDefinition.Initialize()` is called, so any startup logic can be handled in the constructor, or offloaded to a member class that is created in the constructor.

If `TransportInfrastructure` implements `IDisposable`, `Dispose()` will be called when the endpoint stops, so shutdown logic can be handled there.

This is how the RabbitMQ transport currently handles broker connections, via the `ConnectionManager` class.

SzymonPobiega commented:
I don't think we should depend on the fact that container calls `Dispose` on the `TransportInfrastructure`. In other places we prefer an explicit `Stop`. Also `IDisposable` does not play well with `async` stuff. Initialize is also non-async.

andreasohlund commented on src/NServiceBus.Core/StartableEndpoint.cs:
```diff
@@ -31,6 +31,7 @@ public StartableEndpoint(SettingsHolder settings, IBuilder builder, FeatureActiv
             DetectThrottlingConfig();
 
             var pipelineCache = new PipelineCache(builder, settings);
+            await settings.Get<TransportInfrastructure>().Start().ConfigureAwait(false);
```
Getting it from the settings seems hacky, how about we pass it in the ctor instead?

andreasohlund commented on src/NServiceBus.Core/Unicast/RunningEndpointInstance.cs:
```diff
@@ -42,6 +43,7 @@ public RunningEndpointInstance(SettingsHolder settings, IBuilder builder, Pipeli
                 await pipelineCollection.Stop().ConfigureAwait(false);
                 await featureRunner.Stop(messageSession).ConfigureAwait(false);
                 await startAndStoppablesRunner.Stop(messageSession).ConfigureAwait(false);
+                await settings.Get<TransportInfrastructure>().Stop().ConfigureAwait(false);
```
Getting it from the settings seems hacky, how about we pass it in the ctor instead?

andreasohlund commented:
I'm with @SzymonPobiega relying on IDispose seems to always come back to bite us?

This will affect all the downstreams so we need to line up WIP pulls for them?

danielmarbach commented:
I would like to see the use case why this was built linked in the PR here.

SzymonPobiega commented:
@andreasohlund I've made passing of `TransportInfrastructure` explicit, rebased and squashed. The new `Start` and `Send` methods are `virtual` so need need to update in sync. On the other hand, the RabbitMQ code might be refactored to use it.

SzymonPobiega commented:
@danielmarbach I've updated the description.

andreasohlund commented on src/NServiceBus.Core/Transports/TransportInfrastructure.cs:
```diff
@@ -65,5 +66,21 @@ public virtual string MakeCanonicalForm(string transportAddress)
         }
+
+        /// <summary>
+        /// Performs any action required to warm up the transport infrastructure before starting an edpoint.
```
Typo

andreasohlund commented on src/NServiceBus.Core/Transports/TransportInfrastructure.cs:
```diff
@@ -65,5 +66,21 @@ public virtual string MakeCanonicalForm(string transportAddress)
+        }
+
+        /// <summary>
+        /// Performs any action required to cool down the transport infrastructure after stopping an endpoint.
```
Is "after stopping" correct? 

Isn't it more like "Allows the transport to perform cool down actions when the endpoint is stopping"?

SzymonPobiega commented:
@andreasohlund I fixed the typos, rebased and squashed.

andreasohlund commented:
@Particular/transport-maintainers thoughts on this? would this makes sense for our existing transports?

tmasternak commented:
@SzymonPobiega From `SqlServer` perspective `Start/Stop` is not needed.  I think that now the `TransportInfrastructue` gets non trivial from `Core` perspective. It has `Cotr`, `Initialize`, `Start` and `Stop`. 

The life-cycle gets nontrivial. Why not put `Start/Stop` on Send/Receive/Subscription Infrastructures?

SzymonPobiega commented:
I see you point about complexity. When looking at `IInitializableEndpoint`, the `Initialize` method returns a `Task` but it does not do any `async` calls. I believe the initial idea here was to execute the pre-start checks when initializing the endpoint. Then we said we will drop the checks for V6 but left the API async-enabled.

On the other hand, the pre-startup checks of the transport are executed in the `Start` phase (rather than `Initialize`) which I believe is wrong.

Here's what I suggest:
 * Make the `TransportDefinition`'s `Initialize` method `async`
 * Remove the pre-startup checks from `ConfigureSendInfrastructure` and `ConfigureReceiveInfrastructure`
 * Move the checking code to that `Initialize` method
 * Rename `Stop` to `TearDown` to avoid the suggestion that we expect any active thread of execution running inside the `TransportInfrastructure`

tmasternak commented:
:+1:

bording commented:
> I don't think we should depend on the fact that container calls Dispose on the TransportInfrastructure

Just wanted to point out that we're not using the container to call Dispose anymore. We added a [Clear](https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Settings/SettingsHolder.cs#L312) method to `SettingsHolder` and that gets called when the endpoint is stopping.

bording commented:
Before making any changes to the life cycle of `TransportInfrastructure`, it would be a good idea to get feedback from @yvesgoeleven to confirm that doesn't cause issues with ASB.

SzymonPobiega commented:
OK, waiting for @yvesgoeleven then.

SzymonPobiega commented:
I realized that my previous proposal is not doable because the *initialize* phase is executed before the features' setup code so settings are not set yet (e.g. there is no `LocalAddress` value).

Because of that I think that for the V6 release we should add `Start` and `Stop` methods as defined in this PR. It would not be ideal but at least it would allow to implement transport that are more complex than MSMQ. Here's a summary of concerns (that are solvable by this PR):
 * Initializing and disposing of of the connection (or connection pool)
 * Starting and stopping a subscription manager (if it requires an active thread of execution -- polling)
 * Starting and stopping the timeout infrastructure (if it requires an active thread of execution --polling)

andreasohlund commented:
@yvesgoeleven thoughts?

@Particular/nservicebus-maintainers since this is non breaking from a transport perspective and does solve moves to a more explicit lifecycle I'm leaning towards pulling this in? 

@SzymonPobiega is there any transport doco we need to add this to?

timbussmann commented:
I think the `Initialize` `Start` `Stop` `Dispose` lifecycle is a bit over-extensive but if it really solves major problems I'm not against it.

danielmarbach commented:
I'm ok as long as the @Particular/transport-maintainers give their :+1: as well

andreasohlund commented:
Should we raise an issue to kill `Dispose` (since we have `Stop`) in vNext?

timbussmann commented:
> Should we raise an issue to kill Dispose (since we have Stop) in vNext?

As long as we manage the lifecycle of those objects it seems to be fine to remove `IDisposable`.

SzymonPobiega commented:
What do you guys mean by killing `Dispose`?

tmasternak commented:
From SqlServer perspective it's :+1:

danielmarbach commented:
@yvesgoeleven any famous last words before we merge?

yvesgoeleven commented:
@danielmarbach looks like my email overload has hidden this from me, will review

yvesgoeleven commented:
At first sight it doesn't seem to impact ASQ or ASB, unless the responsibility of stopping the pump has to move down, but that's not the case right?

If not  :+1:
